File: examples/urlshortener/common-infra/src/main/java/org/myorg/urls/BaseStack.java
package org.myorg.urls;

import software.amazon.awscdk.Stack;
import software.constructs.Construct;

public class BaseStack extends Stack {
    private final ServiceProps serviceProps;

    public BaseStack(Construct scope, ServiceProps serviceProps) {
        super(scope, serviceProps.service() + "-" + serviceProps.stage());
        this.serviceProps = serviceProps;
    }

    public String org() { return serviceProps.org(); }
    public String subsys() { return serviceProps.subsys(); }
    public String service() { return serviceProps.service(); }
    public Stage stage() { return serviceProps.stage(); }
    public String regionName() { return serviceProps.region(); }
}

File: factory-example/common-stack/src/main/kotlin/org/myorg/sertrac/BaseStack.kt
package org.myorg.sertrac

import software.amazon.awscdk.Stack
import software.constructs.Construct

enum class Stage(val id: String) {
    DEV("dev"),
    QA("qa"),
    PROD("prod");

    override fun toString(): String = id

    companion object {
        fun parse(value: String): org.myorg.sertrac.Stage =
            entries.firstOrNull { it.id.equals(value, ignoreCase = true) }
                ?: throw IllegalArgumentException("Unknown stage: $value")
    }
}

data class ServiceProps(val org : String, val subsys: String, val service: String, val stage: Stage, val region : String) {
}

open class BaseStack(scope: Construct, val serviceProps: ServiceProps) :
    Stack(scope, "${serviceProps.service}-${serviceProps.stage}") {

    public fun org() = serviceProps.org
    public fun subsys() = serviceProps.subsys
    public fun service() = serviceProps.service
    public fun stage() = serviceProps.stage
    public fun regionName() = serviceProps.region

}